ckanext/dia/commands.py

```diff
@@ -69,22 +57,55 @@ def cleanup_datastore_once(self, start_offset):
             sys.exit(1)
 
         # query datastore to get all resources from the _table_metadata
-        resource_id_list = []
-        total_tested = 0
-        for offset in itertools.count(start=start_offset, step=100):
-            print "Load metadata records from datastore (offset: %s)" % offset
-            record_list, tested = self._get_datastore_table_page(context, offset)  # noqa
-            total_tested += tested
-            resource_id_list.extend(record_list)
-            if not tested > 0:
-                break
-            if len(resource_id_list) > 250:
+        print("Loading metadata records from datastore")
+        all_datastore_tables = self._get_all_datastore_tables(context)
+
+        chunk_size = 100
+        total_deleted = 0
+        failed_deletes = []
+        for offset in itertools.count(start=0, step=chunk_size):
+            end_chunk_index = offset + chunk_size
+            datastore_tables_chunk = all_datastore_tables[offset:end_chunk_index]
+            orphan_list = self._find_orphaned_datastore_tables(context, datastore_tables_chunk)
+            if len(orphan_list) > 0:
                 # Run a small chunk of the dataset to avoid locking up the
                 # database for a *really* long time(read: until postgres is
                 # restarted)
+                delete_count, error_list = self._delete_orphans(context, orphan_list)
+                total_deleted += delete_count
+                failed_deletes.extend(error_list)
+                print('Deleted in this chunk: {}'.format(delete_count))
+                print('Errored in this chunk: {}'.format(len(error_list)))
+
+            if end_chunk_index >= len(all_datastore_tables):
                 break
 
-        print('Total resources missing but referenced in datastore: {}'.format(len(resource_id_list)))
+            print('{} tables checked, {} more to go...'.format(end_chunk_index, len(all_datastore_tables) - end_chunk_index))
+            time.sleep(1)
+
+        updated_table_count = len(self._get_all_datastore_tables(context))
+
+        print('Cleanup complete!')
+        print('Total datastore tables before: {} and after: {}'.format(len(all_datastore_tables), updated_table_count))
+        print('Total orphaned datastore tables deleted: {}'.format(total_deleted))
+        print('Total errors when attempting deletion: {}'.format(len(failed_deletes)))
+        if len(failed_deletes) > 0 and len(failed_deletes) < 10: # don't print if too many
+            print('Datastore table names that failed to delete: {}'.format(failed_deletes))
+
+    def _get_all_datastore_tables(self, context):
+        result_set = logic.get_action('datastore_search')(
+            context,
+            {
+                'resource_id': '_table_metadata',
+                'limit': 50000, # arbitrarily high, but default limit is 100
```

Comment:
I suspect the query would time out before 50000 rows anyway.

Reply:
Is it worth having the ability to configure the limit?

Reply:
pass it in as argument to the script or something? I think the only reason you'd want to alter it would be to make it higher, and the timeout of 1 minute hardcoded into the datastore postgres adapter would likely be breached at that point anyway.